command/s3_push.go

```diff
@@ -0,0 +1,123 @@
+}
+
+func s3PushFactory() Command { return &s3Push{} }
+
+func (*s3Push) Name() string {
+	return "s3.push"
+}
+
+func (c *s3Push) ParseParams(params map[string]interface{}) error {
+	if err := mapstructure.Decode(params, c); err != nil {
+		return errors.Wrapf(err, "error decoding %s params", c.Name())
+	}
+	return nil
+}
+
+func (c *s3Push) Execute(ctx context.Context, comm client.Communicator, logger client.LoggerProducer, conf *model.TaskConfig) error {
+	httpClient := util.GetHTTPClient()
+	defer util.PutHTTPClient(httpClient)
+
+	if !c.shouldRunOnBuildVariant(conf.BuildVariant.Name) {
+		logger.Task().Infof("Skipping s3.push for task directory for build variant '%s'", conf.BuildVariant.Name)
+		return nil
+	}
+
+	if err := c.createBucket(httpClient, conf); err != nil {
+		return errors.Wrap(err, "could not set up S3 task bucket")
+	}
+
+	if err := c.bucket.Check(ctx); err != nil {
+		return errors.Wrap(err, "could not find S3 task bucket")
+	}
+
+	wd, err := conf.GetWorkingDirectory("")
+	if err != nil {
+		return errors.Wrap(err, "could not get task working directory")
+	}
+	putMsg := "Pushing task directory files into S3"
+	if c.ExcludeFilter != "" {
+		putMsg += ", excluding files matching filter " + c.ExcludeFilter
+	}
+	logger.Task().Infof(putMsg)
+	if err := c.bucket.Push(ctx, pail.SyncOptions{
+		Local:   wd,
+		Remote:  s3TaskRemotePath(conf),
+		Exclude: c.ExcludeFilter,
+	}); err != nil {
+		return errors.Wrap(err, "error pushing task data to S3")
+	}
+
+	return nil
+}
+
+func (c *s3Push) expandParams(conf *model.TaskConfig) error {
+	if err := util.ExpandValues(c, conf.Expansions); err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (c *s3Push) shouldRunOnBuildVariant(bv string) bool {
+	return util.StringSliceContains(c.BuildVariants, bv)
+}
+
+func (c *s3Push) createBucket(client *http.Client, conf *model.TaskConfig) error {
+	if c.bucket != nil {
+		return nil
+	}
+	opts := pail.S3Options{
+		Credentials: pail.CreateAWSCredentials(conf.S3Data.Key, conf.S3Data.Secret, ""),
+		Region:      endpoints.UsEast1RegionID,
+		Name:        conf.S3Data.Bucket,
+		MaxRetries:  10,
+		Permissions: pail.S3PermissionsPrivate,
+	}
+	bucket, err := pail.NewS3MultiPartBucketWithHTTPClient(client, opts)
+	if err != nil {
+		return errors.Wrap(err, "could not create bucket")
+	}
+	bucket = pail.NewParallelSyncBucket(pail.ParallelBucketOptions{
+		Workers:      runtime.NumCPU(),
+		DeleteOnSync: true,
+	}, bucket)
+	c.bucket = bucket
+	return nil
+}
+
+// s3TaskRemotePath returns the path for the latest s3 push for this task.
+func s3TaskRemotePath(conf *model.TaskConfig) string {
+	return path.Join(s3TaskRemotePathBase(conf), "latest")
+}
+
+// s3TaskPrefixBase is a helper that returns the base s3 path in S3, which is a
+// path path unique to a task.
+func s3TaskRemotePathBase(conf *model.TaskConfig) string {
+	return path.Join(conf.ProjectRef.Identifier, conf.Task.Version, conf.Task.BuildVariant, conf.Task.DisplayName)
```

Comment:
`path.Join` is only a problem when multiple slashes are involved (e.g. in a URL starting with `https://`) because it assumes you're giving it the path portion of the URL after the domain name. It's not a problem in this case, but I removed it since we seem to have low amounts of trust in it.

I changed it to a TaskConfig method.